src/main/xerus/monstercat/api/Playlist.kt

```diff
@@ -86,4 +101,181 @@ object Playlist {
 			else add(track)
 		}
 	}
+}
+
+object PlaylistManager {
+	suspend fun loadPlaylist(apiConnection: APIConnection){
+		val tracks = apiConnection.getTracks()
+		tracks?.forEachIndexed { index, track ->
+			val found = Cache.getTracks().find {
+				it.id == track.id
+			}
+			if (found != null)
+				tracks[index] = found
+			else
+				tracks.removeAt(index)
+		}
+		if (tracks != null && tracks.isNotEmpty()) {
+			if (Player.player?.status != MediaPlayer.Status.DISPOSED)
+				Player.reset()
+			Playlist.setTracks(tracks)
+		}
+	}
+
+	/** Opens the playlist manager dialog
+	 * Allows to load, save, and manage playlists stored on Monstercat.com
+	 */
+	fun playlistDialog(){
+		val connection = APIConnection("api", "playlist").fields(ConnectPlaylist::class)
+
+		val parent = VBox()
+		val stage = App.stage.createStage("Monstercat.com Playlists", parent)
+		stage.initModality(Modality.WINDOW_MODAL)
+		val connectTable = TableView<ConnectPlaylist>()
+
+		// Common playlist functions
+		fun load(runAfter: () -> Unit = {}) {
+			if (connectTable.selectionModel.selectedItem != null) {
+				GlobalScope.async {
+					loadPlaylist(APIConnection("api", "playlist", connectTable.selectionModel.selectedItem.id, "tracks"))
+					onFx { runAfter.invoke() }
+				}
+			}
+		}
+		fun loadUrl(runAfter: () -> Unit = {}) {
```

Comment:
I don't think using this `runAfter` pattern is good. You should rather return an asynchronous value.

Reply:
Can you please explain a bit more ? I have some ideas but I don't think it's better, and it's probably not what you're asking for ^^

Reply:
If you use `async` instead of `launch` you can return a value and then attach the thing you want to run after it to that.

Reply:
Can't seem to make it work with Async. This is my code : https://gist.github.com/defvs/99cbaf6a0aa79b80020eb7d7d5139cd4

Reply:
The code in my invokeOnCompletion is not ran. No idea how to do it otherwise.

Reply:
1. If you use `invokeOnCompletion` you don't even have to use async. Async is only necessary if you return a value. So you can go back to launch ;)
2. Have you checked that the code in the coroutine finishes correctly?

Reply:
I'll try with this, thanks.

Reply:
Okay I've replaced with launch and Job. Some of my jobs invokeOnComplete are still not invoked, and some seems to work just fine ?

Reply:
Nevermind i'm just dumb :P

Reply:
The fact is that I set the job variable that I return in the onClick of my "OK" button, so it always returns null because it still isn't clicked.

Reply:
Is that good ? I mean the indent at the very bottom of the file; it's a lot of curly braces tho.cd56c649088e892ef913e08226067e0767312aa2

Reply:
@Xerus2000

Reply:
No, that's too much - you have my code style now, why don't you just reformat?

Reply:
5382a924d4972475d5f8e9add73450f808258390